### PR TITLE
WIP: Add Aggregation Feature to Flow

### PR DESCRIPTION
**Pull request by kidrahahjo**
## Description
Implementation of helper class for aggregation, `Aggregate`.
This class includes features like:
1.  _Aggregating by numbers_
2. _Aggregating by a condition/statepoint-param_ 
3. _Sorting by a standpoint parameter_
4. _Filtering jobs using the `select` parameter_

## Motivation and Context

The current [PR](https://github.com/glotzerlab/signac-flow/pull/289) for aggregation is too big. Now that PR will track my project and I'll be splitting the work into several small PRs.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

File: flow/aggregate.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from collections.abc import Iterable
+from itertools import groupby
+from itertools import zip_longest
+from functools import partial
+
+
+class _aggregate:
+    """Decorator for operation functions that needs to be aggregated.
```
[the remark this answers is not in the document]
Comment (bdice) [reply flow/aggregate.py:11]: *are*, because "functions" is plural.
```suggestion
    """Decorator for operation functions that are to be aggregated.
```

File: flow/aggregate.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
+# Copyright (c) 2018 The Regents of the University of Michigan
```
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply flow/aggregate.py:1]: Oh yes, sorry.

File: flow/aggregate.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+import signac
+from signac.contrib.project import Project, JobsCursor
+
+
+class aggregate:
+    """Decorator for operation functions that are to be aggregated.
+
+    If this class is used for aggregation then by-default, if the aggregator
+    parameter is not passed, an aggregate of all jobs will be created.
+
```
Comment (atravitz) [inline flow/aggregate.py:19]:
```suggestion
    By default, if the aggregator parameter is not passed, 
    an aggregate of all jobs will be created.

```

File: flow/aggregate.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+from copy import deepcopy
+from itertools import groupby
+from itertools import zip_longest
+from functools import partial
+
+import signac
+from signac.contrib.project import Project, JobsCursor
+
+
+class aggregate:
```
Comment (atravitz) [inline flow/aggregate.py:14]:
```suggestion
class Aggregate:
```

should this be capitalized because it's a class?

File: flow/aggregate.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        example_aggregate = aggregate()
+        @example_aggregate
+        @FlowProject.operation
+        def foo(*jobs):
+            print(len(jobs))
+
+    :param aggregator:
+        Information on how to aggregate jobs. Takes in a list of
+        jobs and can return or yield lists or single jobs instead.
+        The default behavior is creating an aggregate of all jobs.
```
Comment (atravitz) [inline flow/aggregate.py:31]:
```suggestion
        Information describing how to aggregate jobs. Takes in a list of
        jobs and can return or yield lists or single jobs instead.
        The default behavior is creating an aggregate of all jobs.
```

File: flow/aggregate.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    :type aggregator:
+        callable
+    :param sort:
+        Before aggregating, sort the jobs given by a statepoint parameter.
+        The default value is None.
```
Comment (atravitz) [inline flow/aggregate.py:36]:
```suggestion
        Before aggregating, sort the jobs by a given state point parameter.
        The default value is None.
```

File: flow/aggregate.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    :type sort:
+        str or NoneType
+    :param reverse:
+        States if the jobs are to be sorted in reverse order.
+        The default value is False.
+    :type reverse:
+        bool
+    :param select:
+        Condition for filtering jobs. This operates on a single job.
```
Comment (atravitz) [inline flow/aggregate.py:45]:
```suggestion
        Condition for filtering jobs.
```

Why is it important to specify that this operates on a single job?

Comment (kidrahahjo) [reply flow/aggregate.py:45]: Actually the point I was trying to make here is the sequence in which we currently go.
1. select
2. sort
3. aggregate

The `select` parameter should not be something like
```
def _select(job1, job2):
    return job1 > job2
```

File: flow/aggregate.py (hunk 5 above)
Comment (b-butler) [reply flow/aggregate.py:31]:
```suggestion
        Information describing how to aggregate jobs. Is a callable that
        takes in a list of jobs and can return or yield subsets of jobs as list
         or single jobs. The default behavior is creating a single aggregate
         of all jobs.
```
Also, I would recommend that the function should always return list or tuples of subsets of jobs. The yielded or returned value should be an iterable.

File: flow/aggregate.py (hunk 6 above)
Comment (b-butler) [reply flow/aggregate.py:36]: Rather than state the default value, we should state the default behavior here. An user won't necessarily know what `None` means in this context (i.e. no sorting).

File: flow/aggregate.py (hunk 7 above)
Comment (b-butler) [reply flow/aggregate.py:45]:
```suggestion
        Condition for filtering individual jobs. This is passed as the callable argument
        to `filter`.
```

File: flow/aggregate.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        The default value is None
```
Comment (b-butler) [inline flow/aggregate.py:46]: Here as well don't state the default value, state the default behavior.

File: flow/aggregate.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        callable or NoneType
+    """
+
+    def __init__(self, aggregator=None, sort=None, reverse=False, select=None):
+        if aggregator is None:
+            def aggregator(jobs):
+                return jobs
+
+        if not callable(aggregator):
+            raise TypeError("Expected callable aggregator function, got {}"
```
Comment (b-butler) [inline flow/aggregate.py:57]:
```suggestion
            raise TypeError("Expected callable for aggregator, got {}"
```

File: flow/aggregate.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                            "".format(type(aggregator)))
+
+        if sort is not None and not isinstance(sort, str):
+            raise TypeError("Expected string sort parameter, got {}".format(type(sort)))
+
+        if not isinstance(reverse, bool):
+            raise TypeError("Expected bool reverse parameter got {}".format(type(reverse)))
```
Comment (b-butler) [inline flow/aggregate.py:64]: Here I would be lenient we could just do `reverse = bool(reverse)`, so 0 or 1 would work.

File: flow/aggregate.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+        if select is not None and not callable(select):
+            raise TypeError("Expected callable select parameter, got {}".format(type(select)))
```
Comment (b-butler) [inline flow/aggregate.py:67]:
```suggestion
            raise TypeError("Expected callable for select, got {}".format(type(select)))
```

File: flow/aggregate.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+        elif isinstance(key, Iterable):
+            keys = list(key)
+
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                def keyfunction(job):
+                    return [job.sp.get(key, default) for key in keys]
```
Comment (b-butler) [inline flow/aggregate.py:109]: We should allow for `default` to be a list of `len(keys)` here.

File: flow/aggregate.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator, sort, reverse, select)
+
+    def __call__(self, obj=None):
+        if callable(obj):
+            setattr(obj, '_flow_aggregate', self)
+            return obj
+        elif isinstance(obj, (list, Project, JobsCursor)):
```
Comment (b-butler) [inline flow/aggregate.py:128]: This conditional makes me thing we should be dealing with two classes, one for registering the aggregates and another for performing the logic associated with them.

File: flow/aggregate.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+            if not len(aggregated_jobs):
+                return []
+            for i, job in enumerate(aggregated_jobs[-1]):
+                if job is None:
+                    del aggregated_jobs[-1][i:]
+                    break
+            return aggregated_jobs
+        elif isinstance(obj, signac.contrib.job.Job):
+            # For backwards compatibility
+            return [obj]
```
Comment (b-butler) [inline flow/aggregate.py:145]: Why is this needed exactly? Ideally, we would prevent this from ever being called on by a single job internally.

File: flow/aggregate.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        else:
+            raise TypeError('Invalid argument passed while calling '
+                            'the aggregate instance. Expected a callable '
+                            'or an Iterable of Jobs, got {}.'.format(type(obj)))
+
+    def _create_nested_aggregate_list(self, aggregated_jobs):
```
Comment (b-butler) [inline flow/aggregate.py:151]: This function is complicated quite a bit by allowing for iterators that return single jobs. I think we should not allow that behavior.

**Comment by kidrahahjo** [comment]
For now, I'm going to remove the review request from @b-butler @atravitz @csadorf because I'm converting this to a draft.
I will re-request my review after completing this PR

**Comment by kidrahahjo** [comment]
This is being moved to #336
